keps/sig-auth/4193-tokenrequest-requester-info-enhancements/README.md

```diff
@@ -0,0 +1,616 @@
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+* `ServiceAccountTokenJTI` feature flag will toggle including JTI information in tokens, as well as recording JTIs in the audit log / the SA user info.
+* `ServiceAccountTokenNodeInfo` feature flag will toggle including node info in tokens.
+
+Both of these feature flags can be disabled without any unexpected adverse affects or coordination required.
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `ServiceAccountTokenJTI`
+  - Components depending on the feature gate: kube-apiserver
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `EmbedServiceAccountRequesterInfo`
+  - Components depending on the feature gate: kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+Enabling the feature gate will cause additional information to be stored/persisted into service account JWTs, as well
+as new audit annotations being recorded in the audit log. This is all purely additive, so no changes to existing
+features, schemas or fields are expected.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Future tokens will then not embed this information. Any existing issued tokens **will** still have this
+information embedded however.
+
+If these fields are deemed to be problematic for other systems interpreting these tokens, users will need to re-issue
+these tokens before presenting them elsewhere.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Future tokens will once again include this information/no adverse effects.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes (as noted above in the test plan)
+
+### Rollout, Upgrade and Rollback Planning
+
+Rolling this out will be done by enabling the feature flag on all control plane hosts.
+This should not have any issues/affect during upgrades.
+Rollback is done by removing/disabling the feature gate.
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+If somehow the tokens issued during the time where the feature was enabled were to be rejected, these tokens would
+need to be re-requested.
+
+This isn't so much a failure of the rollback, rather it would be a bug in the actual feature (hence feature gating).
+
+During a skew where hosts have different enablement, users would get tokens with varying content if they made multiple
```

Comment:
how would an old server with the featuregate off verify a token bound to a node object?

Reply:
Good catch.

Options I can think of:
1. Server without feature gate enabled ignores node-bound aspect, just validates the token signature and service account existence
    * This makes the tokens *less* secure when the feature gate is disabled... I don't like this option
2. Server without feature gate rejects node-bound tokens.
    * This would mean node-bound tokens shouldn't be requested and relied on until upgrade to enabled-state was complete and rollback safe.
3. Server without feature gate does not allow creation of node-bound tokens but continues to accept/validate them if presented. Requires running the node informer.
    * Since ~all deployments I'm aware of run the node informer anyway as part of the node authorizer, this doesn't seem bad to me.
    * 3a: If we wanted to give very fine-grained control over this, we could gate issuance and acceptance of explicitly node-scoped tokens separately, graduate acceptance in one release, then issuance in the next.

I'd lean toward 3 or 3a.

Reply:
I'm fine with 3a. I think it's better than being less secure.

Reply:
👍 I will add an extra feature gate to the proposal, and also note that the rollout plan will require an additional release for graduation.

Reply:
Updated to reflect 3a - please take a look & let me know if we need any changes, we *may* want to create a 4th feature gate to control whether we enable the explicitly node scope token *verification* behaviour too, unless we are happy to not have this behaviour feature gated.